packages/gatsby/src/utils/jobs-manager.js

```diff
@@ -0,0 +1,261 @@
+  // currently this needs to be sync as we could miss jobs to have been scheduled and
+  // are still processing their hashes
+  const inputPathsWithContentDigest = inputPaths.map(path => {
+    return {
+      path: convertPathsToAbsolute(path),
+      contentDigest: createFileHash(path),
+    }
+  })
+
+  /** @type {InternalJob} */
+  const internalJob = {
+    id: uuid(),
+    name,
+    contentDigest: ``,
+    inputPaths: inputPathsWithContentDigest,
+    outputDir: convertPathsToAbsolute(outputDir),
+    args,
+    plugin: {
+      name: plugin.name,
+      version: plugin.version,
+      resolve: plugin.resolve,
+      isLocal: !plugin.resolve.includes(`/node_modules/`),
+    },
+  }
+
+  // generate a contentDigest based on all parameters including file content
+  internalJob.contentDigest = createContentDigest({
+    name: job.name,
+    inputPaths: internalJob.inputPaths.map(
+      inputPath => inputPath.contentDigest
+    ),
+    outputDir: internalJob.outputDir,
+    args: internalJob.args,
+    plugin: internalJob.plugin,
+  })
+
+  return internalJob
+}
+
+/**
+ * Creates a job
+ *
+ * @param {InternalJob} job
+ * @return {Promise<object>}
+ */
+exports.enqueueJob = async job => {
+  // When we already have a job that's executing, return the same promise.
+  // we have another check in our createJobV2 action to return jobs that have been done in a previous gatsby run
+  if (jobsInProcess.has(job.contentDigest)) {
+    return jobsInProcess.get(job.contentDigest).deferred.promise
+  }
+
+  // Bump active jobs
+  activeJobs++
+  if (!activityForJobs) {
+    activityForJobs = reporter.phantomActivity(`Running jobs v2`)
+    activityForJobs.start()
+  }
+
+  const deferred = pDefer()
+  jobsInProcess.set(job.contentDigest, {
+    id: job.id,
+    deferred,
+  })
+
+  try {
+    const result = await runJob(job)
+    // this check is to keep our worker results consistent for cloud
+    if (result != null && !_.isPlainObject(result)) {
+      throw new Error(
+        `Result of a worker should be an object, type of "${typeof result}" was given`
+      )
+    }
+    deferred.resolve(result)
+  } catch (err) {
+    deferred.reject(err)
+  } finally {
+    // when all jobs are done we end the activity
+    if (--activeJobs === 0) {
+      activityForJobs.end()
+      // eslint-disable-next-line require-atomic-updates
+      activityForJobs = null
+    }
+  }
+
+  return deferred.promise
+}
+
+/**
+ * Remove a job from our inProgressQueue to reduce memory usage
+ *
+ * @param {string} contentDigest
+ */
+exports.removeInProgressJob = contentDigest => {
+  jobsInProcess.delete(contentDigest)
+}
+
+/**
+ * Wait for all processing jobs to have finished
+ *
+ * @return {Promise<void>}
+ */
+exports.waitUntilAllJobsComplete = () => {
+  const jobsPromises = []
+  jobsInProcess.forEach(({ deferred }) => jobsPromises.push(deferred.promise))
+
+  return Promise.all(jobsPromises).then(() => {})
```

Comment:
This might not scale very well. `jobsInProcess` can have thousands of promises. Would it be possible to instead have counters in jobs and jobsv2 that will emit single `drain` (or `done`) event (emit this in endJob action if there is no more jobs to do)?

Reply:
(This might be follow up work)

Reply:
(there is already check like what I have in mind in `enqueueJob` related to phantom activity we are creating)